lib/kube/proxy/cluster_details.go

```diff
@@ -89,21 +113,69 @@ func newClusterDetails(ctx context.Context, cfg clusterDetailsConfig) (*kubeDeta
 		go dynLabels.Start()
 	}
 
-	return &kubeDetails{
-		kubeCreds:     creds,
-		dynamicLabels: dynLabels,
-		kubeCluster:   cfg.cluster,
-	}, nil
+	// Create the codec factory and the list of supported types for RBAC.
+	codecFactory, rbacSupportedTypes, err := newClusterSchemaBuilder(creds.getKubeClient())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	k := &kubeDetails{
+		kubeCreds:          creds,
+		dynamicLabels:      dynLabels,
+		kubeCluster:        cfg.cluster,
+		kubeCodecs:         codecFactory,
+		rbacSupportedTypes: rbacSupportedTypes,
+		rwMu:               &sync.RWMutex{},
+		closeC:             make(chan struct{}),
+	}
+
+	// Start the periodic update of the codec factory and the list of supported types for RBAC.
+	go func() {
+		defer func() {
+			k.closeC <- struct{}{}
+		}()
+		ticker := cfg.clock.NewTicker(5 * time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-k.closeC:
+				return
+			case <-ticker.Chan():
+				codecFactory, rbacSupportedTypes, err := newClusterSchemaBuilder(creds.getKubeClient())
+				if err != nil {
+					cfg.log.WithError(err).Error("Failed to update cluster schema")
+					continue
+				}
+
+				k.rwMu.Lock()
+				k.kubeCodecs = codecFactory
+				k.rbacSupportedTypes = rbacSupportedTypes
+				k.rwMu.Unlock()
+			}
+		}
+	}()
+	return k, nil
 }
 
 func (k *kubeDetails) Close() {
+	// send a close signal and wait for the close to finish.
+	k.closeC <- struct{}{}
+	<-k.closeC
```

Comment:
I changed the approach but didn't have a race because the channel isn't buffered.

Since the channel wasn't buffered, the consumer and producer must sync in order to send the signal, and the consumer blocks until this operation happens.

Reply:
Much better, thanks.